### ISSUE TITLE
Maya Tool - Line Up Toolset

### ISSUE THREAD

**Issue by david-cattermole** [david-cattermole/mayaMatchMoveSolver#121]
### Feature
Create toolset for fast camera line ups without LIDAR or survey, and focusing on line constraints or geometrically known information of the set. These tools are wrappers around mmSolver, designed to create mmSolver setups easily, without users needing intricate knowledge of mmSolver.

This feature is inspired by [fSpy](https://fspy.io/).

This is a list of tools that may be helpful:
- Create Camera
- Add Survey Point
- Add Line (X Axis)
- Add Line (Y Axis)
- Add Line (Z Axis)
- Add Cube
- Add Plane (XZ Axis)
- Add Horizon Line (Camera Roll)
- Solve Line-Up Frame

All node networks created for the Line Up should be parented under the group node `CameraLineup_GRP`.

#### Create Camera

This tool will create a new default camera, with settings and nodes set up for other Line Up tools.
The camera should be created under the `CameraLineup_GRP` node.

#### Add Survey Point
Allow the user to give the solver a 3D position, as known data.

#### Add Line X/Y/Z Axis
Creates a new rig set up for solving a Line on an axis.

1. Select Camera or click in active viewport.
2. Run tool; A node network is created, parented under `CameraLineup_GRP`.
3. Position Markers along the X/Y/Z plane, as needed.

#### Add Cube
Similar to Add Line tool, but creates 8 Markers by default, the user is expected to position Markers and disable Markers that cannot be seen.

#### Add Plane
Similar to Add Line tool, but creates 1 Marker by default and places it on the XZ plane, with TY locked to zero.

#### Add Horizon Line
Create a Marker Line with bundles connected to an "infinite circle".
This tool should be used especially for solving Camera RZ.

#### Solve Line-Up Frame
Creates a Collection (with a preset name) with all Markers/Attributes and runs a solve on the current frame.

1. Add Camera and Image Plane.
2. Add Lines, Cube, etc.
3. Select all nodes you want to use for the Line Up; only selected nodes are used in the line-up.
4. Run tool; The selected nodes are found, and converted into "preset-logic" for solving a line up with the given information.

Some solving ideas are:
- Move objects in front of the camera (using re-projection).
- Don't always solve everything together, work out a few simple steps that can be used. 

### General Workflow

1. Create new Maya scene
2. Create new Camera
3. Add Image Plane to the Camera.
4. Add a cube; 8 Markers are created, with a 3D cube at origin.
5. Position cube Markers on image.
6. Press "Solve Line-Up Frame" tool.

**Comment by david-cattermole** [comment]
Rather than re-implementing the new solver, it would make sense to use the fSpy command line interface recently introduced in v1.0.3.

https://github.com/stuffmatic/fSpy/releases/tag/v1.0.3

We could install fSpy as part of mmSolver and use the executable directly. However we would need to compile the Linux binary, as it seems fSpy doesn't provide a Linux binary.

**Comment by stuffmatic** [comment]
fSpy Linux builds are distributed as an executable AppImage files. Would that work?

**Comment by david-cattermole** [comment]
Oh, right. I didn't realise .AppImage was for Linux. I'll try it out. 

Thanks @stuffmatic

**Comment by david-cattermole** [comment]
I have implemented a custom Maya node that can calculate the camera translation, rotation and focal length, for images with 2 vanishing points (2 lines creating each vanishing point).

The following modes are supported:
- one point perspective (1 vanishing point) - camera tilt and pan, and translation estimation.
- one point perspective with horizon line  (1 vanishing points + 1 line) - camera tilt, pan and roll, and translation estimation.
- two point perspective (2 vanishing points) - camera tilt, pan and roll, and translation estimation.

The scale cannot be calculated from a single image, however I've added a scale mode so that when the (2D) origin point is on the ground we can scale the camera to ensure the camera height is a certain number. This allows us to use any camera data recorded on-set or a simple guess since a human will often be holding the camera.

There is a tool to create a default setup, however there are a number of limitations to this tool currently, and so I have classed this as "beta". Limitations include:

- No GUI has been created for the tool.

- There are no line visualizations to show the individual lines.

- The camera does not yet update automatically while dragging the 2D points.

- Only undistorted images should be used.

  Straight-lines are assumed to be straight - no lens distortion is
  taken into account.

- Only 2 vanishing points can be used.

  Some images may require 3 vanishing points, if the center of the
  lens ("principal point" or "camera film offsets") is not in the
  center of the image.

  This is fine for most use-cases with (typical) un-cropped images,
  however issues may arise when a Tilt-Shift lens was used for the
  image, or the image has been cropped non-uniformly.

  If you have this problem it is recommended to use "fSpy" to
  calibrate the camera, and "maya_fspy" to import the camera into
  Maya.

https://github.com/stuffmatic/fSpy
https://github.com/JustinPedersen/maya_fspy

Even with these limitations, I'd like to release this tool in the next v0.3.x release of mmSolver, however I will hopefully create some an icon and documentation for the next v0.3.x release.
